### DISCUSSION TITLE
Bind a table or aggrid to a dataframe/ Update table or aggrid

### DISCUSSION THREAD

**Discussion by hoderlump** [zauberzeug/nicegui#1521] [Q&A] [answered]
### Question

Hello,

I am working on this for many hours but I cannot fix it. It seems easy but it is not (for me at least...) :)
I have a Pandas dataframe and I want to show that data in a table or better, in a aggrid.
This is easy.

What I cannot do is update the table/ aggrid.

This is my minimal example:

```python
import pandas as pd
from nicegui import app, ui

df = pd.DataFrame()

async def choose_csv_file():
    files = await app.native.main_window.create_file_dialog(allow_multiple=True, file_types=("CSV file (*.csv)", ))
    global df
    for file in files:
        df = pd.concat([pd.read_csv(file), df])

def update():
    datatable1.clear()
    with datatable1:
        ui.aggrid.from_pandas(df)

with ui.row():
    ui.button('Add csv measurement file', on_click=choose_csv_file).classes("w-40")
    datatable1 = ui.aggrid.from_pandas(df).style('width: max')
    ui.button('Update', on_click=update)

ui.run(native=True, dark=True, reload=False)

```

The aggrid will not update, whatever I do. I I use a table instead, the issue is the same.
The proposed methods as aggrid.update() or ui.update() do not work.
Maybe this is related to issue https://github.com/zauberzeug/nicegui/issues/600?

As CSV, any CSV with comma separated values can be used. I use that one: https://cdn.wsform.com/wp-content/uploads/2020/06/color_web_safe.csv

Thank you very much for your help!

**Comment by falkoschindler** [comment, marked as answer]
Hi @hoderlump!

You're clearing the table (which removes all child elements - but the table has no children, so this operation does nothing) and create a new table _inside_ the existing table (which isn't supposed to render any children, thus you're seeing nothing).

Instead you can put the table inside a container element which you clear and re-populate with a new table:
```py
def update():
    table_container.clear()
    with table_container:
        ui.aggrid.from_pandas(df)

...

with ui.element() as table_container:
    ui.aggrid.from_pandas(df).style('width: max')
```

You just need to figure out the layouting...

**Reply by hoderlump** [reply]
This works well, thanks a lot!
The tailwind layouting of aggrid does not work as expected for me but CSS/.style() this works here:
```python
def update():
    table_container.clear()
    with table_container:
        ui.aggrid.from_pandas(df).style("width: 100%;")
```
...
```python
with ui.element().style("width: calc(100% - 420px);") as table_container:
    ui.aggrid.from_pandas(df).style("width: 100%;")
```

**Comment by stdusr** [comment]
I am facing similar issue, but the solution provided by [falkoschindler](https://github.com/falkoschindler) doesn't work, for the heck of me:

```py
from nicegui import ui
import pandas as pd

with ui.element() as table_container:
    df = pd.read_csv('https://people.sc.fsu.edu/~jburkardt/data/csv/biostats.csv')
    ui.aggrid.from_pandas(df)

ui.run()
```
I just see a grey vertical bar. Without ui.element() it works fine. Could someone correct me, please?

**Reply by hoderlump** [reply]
You need to give the container (and maybe the aggrid itself) a certain size. 
For aggrid, I think only .style() works well, I think. Try this:

```python
with ui.element().classes("w-full") as table_container:
    ui.aggrid.from_pandas(df).style("width: 50%;")
```

I am struggling a lot with the layout, especially with vertical alignments myself right now. It seems, a lot of trickery is needed here and there sometimes.

**Reply by stdusr** [reply]
Works. Thanks a million!
